rcl_action/test/rcl_action/test_goal_handle.cpp

```diff
@@ -0,0 +1,309 @@
+  rcl_ret_t ret = rcl_action_goal_handle_get_info(nullptr, &goal_info_output);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check with invalid goal handle
+  rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+  ret = rcl_action_goal_handle_get_info(&goal_handle, &goal_info_output);
+  EXPECT_EQ(ret, RCL_RET_ACTION_GOAL_HANDLE_INVALID) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check with null goal info
+  ret = rcl_action_goal_handle_init(&goal_handle, &goal_info_input, rcl_get_default_allocator());
+  ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  ret = rcl_action_goal_handle_get_info(&goal_handle, nullptr);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check with valid arguments
+  ret = rcl_action_goal_handle_get_info(&goal_handle, &goal_info_output);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  for (int i = 0; i < 16; ++i) {
+    EXPECT_EQ(goal_info_input.uuid[i], goal_info_output.uuid[i]);
+  }
+  EXPECT_EQ(goal_info_input.stamp.sec, goal_info_output.stamp.sec);
+  EXPECT_EQ(goal_info_input.stamp.nanosec, goal_info_output.stamp.nanosec);
+
+  // Finalize
+  ret = rcl_action_goal_handle_fini(&goal_handle);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+}
+
+TEST(TestGoalHandle, test_goal_handle_update_state_invalid)
+{
+  // Check with null argument
+  rcl_ret_t ret = rcl_action_update_goal_state(nullptr, GOAL_EVENT_EXECUTE);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check with invalid goal handle
+  rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+  ret = rcl_action_update_goal_state(&goal_handle, GOAL_EVENT_NUM_EVENTS);
+  EXPECT_EQ(ret, RCL_RET_ACTION_GOAL_HANDLE_INVALID) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check with invalid goal event
+  rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+  ret = rcl_action_goal_handle_init(&goal_handle, &goal_info, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  ret = rcl_action_update_goal_state(&goal_handle, GOAL_EVENT_NUM_EVENTS);
+  EXPECT_EQ(ret, RCL_RET_ACTION_GOAL_EVENT_INVALID) << rcl_get_error_string().str;
+  rcl_reset_error();
+}
+
+typedef std::vector<std::pair<rcl_action_goal_event_t, rcl_action_goal_state_t>>
+  StateTransitionSequence;
+const std::vector<std::string> event_strs = {
+  "EXECUTE", "CANCEL", "SET_SUCCEEDED", "SET_ABORTED", "SET_CANCELED"};
+
+class TestGoalHandleStateTransitionSequence
+  : public ::testing::TestWithParam<StateTransitionSequence>
+{
+public:
+  static std::string print_sequence_param_name(
+    const testing::TestParamInfo<StateTransitionSequence> & info)
+  {
+    std::string result = "";
+    for (auto event_state : info.param) {
```

Comment:
@jacobperron `auto` -> `const auto &`

Reply:
I'd also be explicit about the type, just for the sake of readability.

Reply:
Fixed in 29f0183